File: backend/market_platform/schemas/news.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field


class EventAttributeOut(BaseModel):
    key: str
    value_text: str | None = None
    value_number: Decimal | None = None
    unit: str | None = None
    source_text: str | None = None
    extraction_method: str
    confidence: Decimal | None = None


class NewsMetrics(BaseModel):
    revenue_yoy_pct: float | None = None
    revenue_qoq_pct: float | None = None
    ebitda_yoy_pct: float | None = None
    pat_yoy_pct: float | None = None
    pat_qoq_pct: float | None = None
    eps_yoy_pct: float | None = None
    ocf_yoy_pct: float | None = None
    revenue_cr: float | None = None
    ebitda_cr: float | None = None
    pat_cr: float | None = None
    order_value_cr: float | None = None
    order_to_revenue_pct: float | None = None
    materiality_band: str | None = None


class NewsEventCard(BaseModel):
    id: int
    event_type: str
    category: str
    category_label: str
    title: str
    what_happened: str
    summary_simple: str | None = None
    impact_score: Decimal | None = None
    confidence: Decimal | None = None
    sentiment: str | None = None
    published_at: datetime | None = None
    price_reaction_pct: Decimal | None = None
    extraction_method: str | None = None
    stock_id: int | None = None
    symbol: str | None = None
    company_name: str | None = None
    exchange: str | None = None
    sector: str | None = None
    last_price: Decimal | None = None
    change_pct: float | None = None
    source_url: str | None = None
    announcement_id: int
    attributes: list[EventAttributeOut] = Field(default_factory=list)
    metrics: NewsMetrics = Field(default_factory=NewsMetrics)
    sector_profile: str = "general"
    sector_kpis: dict[str, float | None] = Field(default_factory=dict)


class NewsListResponse(BaseModel):
    items: list[NewsEventCard]
    total: int
    limit: int
    offset: int


class NewsCategoryCount(BaseModel):
    key: str
    label: str
    count: int


class NewsCategoriesResponse(BaseModel):
    categories: list[NewsCategoryCount]
    total: int


class TrendingStockOut(BaseModel):
    rank: int
    stock_id: int
    symbol: str
    company_name: str
    exchange: str
    event_count: int
    last_price: Decimal | None = None
    change_pct: float | None = None


class TrendingNewsResponse(BaseModel):
    items: list[TrendingStockOut]


class SummarizeResponse(BaseModel):
    id: int
    summary_simple: str | None
    extraction_method: str | None
    llm_used: bool


class NewsRefreshResponse(BaseModel):
    fetched: int
    created: int
    processed: int
    skipped: int
    days: int
    exchanges: list[str]
    used_seed: bool
    live_count: int
    purged_seed: int = 0
    limit: int = 0

File: web/src/types/news.ts
export type NewsCategoryKey =
  | "all"
  | "result"
  | "award_order"
  | "press_release"
  | "monthly_update"
  | "shareholder"
  | "other";

export interface EventAttribute {
  key: string;
  value_text: string | null;
  value_number: string | number | null;
  unit: string | null;
  source_text: string | null;
  extraction_method: string;
  confidence: string | number | null;
}

export interface NewsMetrics {
  revenue_yoy_pct: number | null;
  revenue_qoq_pct: number | null;
  ebitda_yoy_pct: number | null;
  pat_yoy_pct: number | null;
  pat_qoq_pct: number | null;
  eps_yoy_pct: number | null;
  ocf_yoy_pct: number | null;
  revenue_cr: number | null;
  ebitda_cr: number | null;
  pat_cr: number | null;
  order_value_cr: number | null;
  order_to_revenue_pct: number | null;
  materiality_band: string | null;
}

export type SectorProfile = "bank" | "nbfc" | "it" | "pharma" | "general" | string;

export interface NewsEventCard {
  id: number;
  event_type: string;
  category: string;
  category_label: string;
  title: string;
  what_happened: string;
  summary_simple: string | null;
  impact_score: string | number | null;
  confidence: string | number | null;
  sentiment: string | null;
  published_at: string | null;
  price_reaction_pct: string | number | null;
  extraction_method: string | null;
  stock_id: number | null;
  symbol: string | null;
  company_name: string | null;
  exchange: string | null;
  sector?: string | null;
  last_price: string | number | null;
  change_pct: number | null;
  source_url: string | null;
  announcement_id: number;
  attributes: EventAttribute[];
  metrics?: NewsMetrics | null;
  sector_profile?: SectorProfile;
  sector_kpis?: Record<string, number | null>;
}

export interface NewsListResponse {
  items: NewsEventCard[];
  total: number;
  limit: number;
  offset: number;
}

export interface NewsCategoryCount {
  key: string;
  label: string;
  count: number;
}

export interface NewsCategoriesResponse {
  categories: NewsCategoryCount[];
  total: number;
}

export interface TrendingStock {
  rank: number;
  stock_id: number;
  symbol: string;
  company_name: string;
  exchange: string;
  event_count: number;
  last_price: string | number | null;
  change_pct: number | null;
}

export interface TrendingNewsResponse {
  items: TrendingStock[];
}

export interface SummarizeResponse {
  id: number;
  summary_simple: string | null;
  extraction_method: string | null;
  llm_used: boolean;
}

export interface NewsRefreshResponse {
  fetched: number;
  created: number;
  processed: number;
  skipped: number;
  days: number;
  exchanges: string[];
  used_seed: boolean;
  live_count: number;
  purged_seed: number;
  limit: number;
}

export interface NewsQuery {
  category?: string;
  event_type?: string;
  symbol?: string;
  q?: string;
  date_from?: string;
  date_to?: string;
  important_only?: boolean;
  sentiment?: string;
  impact_band?: string;
  sort?: "latest" | "impact";
  limit?: number;
  offset?: number;
}
